py/private/toolchain/repo.bzl

```diff
@@ -0,0 +1,77 @@
+"""Create a repository to hold the toolchains
+
+This follows guidance here:
+https://docs.bazel.build/versions/main/skylark/deploying.html#registering-toolchains
+"
+Note that in order to resolve toolchains in the analysis phase
+Bazel needs to analyze all toolchain targets that are registered.
+Bazel will not need to analyze all targets referenced by toolchain.toolchain attribute.
+If in order to register toolchains you need to perform complex computation in the repository,
+consider splitting the repository with toolchain targets
+from the repository with <LANG>_toolchain targets.
+Former will be always fetched,
+and the latter will only be fetched when user actually needs to build <LANG> code.
+"
+The "complex computation" in our case is simply downloading our pre-built rust binaries.
+This guidance tells us how to avoid that: we put the toolchain targets in the alias repository
+with only the toolchain attribute pointing into the platform-specific repositories.
+"""
+
+load("//py/private/toolchain:tools.bzl", "RUST_BIN_CFG", "TOOLCHAIN_PLATFORMS")
+
+def _toolchains_repo_impl(repository_ctx):
+    build_content = """# Generated by toolchains_repo.bzl
+#
+# These can be registered in the workspace file or passed to --extra_toolchains flag.
+# By default all these toolchains are registered by the py_register_toolchains macro
+# so you don't normally need to interact with these targets.
+
+"""
+    for tool, cfg in RUST_BIN_CFG.items():
+        for [platform, meta] in TOOLCHAIN_PLATFORMS.items():
+            build_content += """
+# Declare a toolchain Bazel will select for running {tool} on the {cfg} platform.
+toolchain(
+    name = "{tool}_{platform}_toolchain",
+    {cfg}_compatible_with = {compatible_with},
+    # Bazel does not follow this attribute during analysis, so the referenced repo
+    # will only be fetched if this toolchain is selected.
+    toolchain = "@{user_repository_name}.{platform}//:{tool}_toolchain",
+    toolchain_type = "@aspect_rules_py//py/private/toolchain:{tool}_toolchain_type",
```

Comment:
Do we need a new toolchain type for exec and target platforms? if any of these tools are invoked at runtime and compile time?

Reply:
Nope, the reason we have two tools is that venv is only used at runtime in a binary, and unpack is only used in an action. Commented that over here https://github.com/aspect-build/rules_py/pull/292/files#diff-d06e56acb93ae623783dbfa6032fe99d955510112ef0fdaf59286b3d3e564ae4R8